### ISSUE TITLE
Enhancement: animations for water

### ISSUE THREAD

**Issue by colonelfazackerley** [andybalaam/rabbit-escape#350]
for #298 

I thought this might be worth a separate issue to the cell mechanics.

I have some GNU Octave code to make sprites.

**Comment by colonelfazackerley** [comment]
![re](https://cloud.githubusercontent.com/assets/9596719/13264317/78ee2682-da65-11e5-9923-9db2dffa6f6f.gif)
This gif is a bit naff, 
- the in-game pngs will have a transparent background.
- they are different sizes, the gif here is cropped to the smallest.

the idea is that these sprites slosh around overlapping and looking vaguely watery.

**Comment by tttppp** [comment]
Looks watery! I wondered if it might be worth having an overlay for a cell when it's lethal to be in? Just a blue mostly transparent square?

Anyway I hope to have something where the water moves downwards soon, so you will be able to test falling water.

How many sho she's do you anticipate having in a cell? The contents is currently an int that goes from 0 to 100 (or 0 to 50 for a ramp).

**Comment by colonelfazackerley** [comment]
Need to experiment. I was thinking a water sprite would be worth 10 of your units. so 10 of them in a full cell.

I am a bit nervous that it will be computationally too expensive.

**Comment by tttppp** [comment]
*"sho she's" is meant to say "sloshes".

Also the blue overlay is meant to be in addition to the sloshes. I was thinking it might be hard for the player to tell whether it was safe to let rabbits march through some water or not.

**Comment by colonelfazackerley** [comment]
Overlays are fine for a pool on a flat surface. I cant think how to make them look good for streams plunging over cliffs. We will just have to try some stuff and see what works.

**Comment by colonelfazackerley** [comment]
I have made a tentative start (does not do anything yet)

https://github.com/tttppp/rabbit-escape/compare/Water...colonelfazackerley:Water?expand=1

**Comment by andybalaam** [comment]
I am enjoying watching how this is going. I must say I can't imagine how it will look yet and wonder whether some hand drawn pngs or svgs might be cool but I'm happy to be convinced.

**Comment by colonelfazackerley** [comment]
Making some slow progress. I still have not got to the point where it will compile. Writing this much code without running something makes me nervous. I guess the answer is to write some unit tests.

**Comment by colonelfazackerley** [comment]
My branch compiles and runs. Yay! 

But no water sprites appear. Boo :(

**Comment by colonelfazackerley** [comment]
Those wiggly worms are water. This obviously needs a lot of work.

![re](https://cloud.githubusercontent.com/assets/9596719/13511996/df0ded36-e18f-11e5-9f91-3a41f33fdf56.gif)

**Comment by andybalaam** [comment]
Excited you have something running!

**Comment by colonelfazackerley** [comment]
I have lots of bugs in the water particle movement. I will have to concede that the water regions need something like a filled polygon with particles on top to give a watery texture. 

To make a cell look full it would take at least 50 water sprites, which is too slow.

Maybe falling water can be sprites alone.

**Comment by tttppp** [comment]
Showing a custom animation for falling water is something I didn't bother coding in (as seen in your animated gif a few comments up). In the python prototype I determined if water was falling by checking if water could flow downwards (i.e. is there a connection downwards, and if so is the connected region full?)

If we wanted to just use sprites, could you add ten worms to a cell and record that as a single sprite? That way we would get fuller looking cells without extra object handling. If this idea makes no sense then don't worry, using a combination of polygons and sprites sounds fine to me.

**Comment by colonelfazackerley** [comment]
![re](https://cloud.githubusercontent.com/assets/9596719/13581904/1d93e39e-e4a1-11e5-8176-1c499566cd94.gif)

needs falling water and something on the slope block.

**Comment by colonelfazackerley** [comment]
I made a branch with the sprites removed.
![re](https://cloud.githubusercontent.com/assets/9596719/13669540/1c73036a-e6bc-11e5-99b5-4172c42fd72b.gif)

i think i should
- tidy this code
- implement tests
- play tttppp's levels and resolve issues
- rebase/rebranch from tttppp/Water and then cherry pick
- merge this into tttppp/Water
- merge into andybalaam/master
- implement particles so the falling water can be seen.

**Comment by tttppp** [comment]
I've been watching the animation avidly over lunch. This looks much better than the previous version. Water mainly builds up from the bottom, which is much more realistic. Here are the problems I see with the polygons at the moment.

1. The initial pipe creates a polygon for one frame.

2. When filling to the second layer, the water level jumps violently.

I understand why both things happen, but I think it doesn't look brilliant. Maybe sprites for falling water will reduce the visibility of these things? I don't have better suggestions either, I just thought it was worth noting the problems.

**Comment by colonelfazackerley** [comment]
I noticed that too. I think I understand the violent jump, but not the pipe filling for one frame (yet).

I will see if I can do something about them. I may play your levels first: there may be other issues.

**Comment by tttppp** [comment]
I think the pipe filling for one frame will be because there's no flow associated with pipes. I assume the code determines whether water is falling based on the flow, and pipes don't update the flow at all (they just add directly to the contents of a cell).  After the first time step then water in the pipe's cell falls, and so relying on the flow becomes useful again.  It might be possible to fix by updating the Pipe code to use the flow in `calcNewState`, rather than setting the contents in `step`.

**Comment by andybalaam** [comment]
Wow yes I'm starting to see how this might work!

**Comment by colonelfazackerley** [comment]
![re](https://cloud.githubusercontent.com/assets/9596719/13708633/9cd4fe00-e7a7-11e5-859d-b315d0ece28e.gif)
I fixed some glitches and introduced others.

**Comment by tttppp** [comment]
That looks playable to me. I like how the water looks like it's sloshing about!

**Comment by colonelfazackerley** [comment]
https://github.com/andybalaam/rabbit-escape/compare/master...colonelfazackerley:Water2

Needs a rebase and tests.

I played Dig Quick, and it was OK.

**Comment by colonelfazackerley** [comment]
clearer diff agaist tttppp/Water

https://github.com/tttppp/rabbit-escape/compare/Water...colonelfazackerley:Water2

**Comment by tttppp** [comment]
I just tried Dig Quick with that branch (ad188f4e36eb882f6e2fce27c81d28004f203ff3), and it's really slow on my machine. I manually deleted all the debug printing, and it was better, but still fairly slow. Is `make rungui` the right way to run the game, or is there some other method to avoid debug processing?

**Comment by colonelfazackerley** [comment]
oh, no, the diff has loads of other stuff in too.

**Comment by colonelfazackerley** [comment]
I have not made the debug stuff configurable. i will remove the calls when it's ready.

**Comment by tttppp** [comment]
The shapes of the water flowing look nice. It shows how near to full a cell is, which is a huge help compared with the static frames in tttppp/Water. I think currently it's probably too slow though. I haven't looked at the code at all, but will do when I get chance. Maybe removing the debugging will be enough?

**Comment by andybalaam** [comment]
I'll give this a go when I get a minute and try to merge tttppp Water as soon as I can.

**Comment by colonelfazackerley** [comment]
![re](https://cloud.githubusercontent.com/assets/9596719/13708633/9cd4fe00-e7a7-11e5-859d-b315d0ece28e.gif)
I fixed some glitches and introduced others.

**Comment by colonelfazackerley** [comment]
I ran a profiler (visualvm - from ubuntu repo - it seems pretty sweet).
![profilersnapshot](https://cloud.githubusercontent.com/assets/9596719/13722879/991a8b6e-e84b-11e5-9533-096ec7b718e2.png)

The problem is drawFilledPoly. that method just calls Graphics.fillPolygon. So the problem is that swing has a poorly implemented method.

- could make a whole bunch of pngs instead of those polygons. should compress well if they are only 2 colours.
- could calculate one polygon, so a single call to fillPolygon covers several cells.

other ideas?

**Comment by colonelfazackerley** [comment]
polygon drawing does not work on android (yet), but i expect it's graphics methods are better...

i did keep as much code in render package as possible, so it should be quick to get it working on android.

**Comment by tttppp** [comment]
Am I reading it right that it drew 4000 polygons in 70ms? Or is the time 70000ms (=70s)? Are rectangles any cheaper to draw than polygons? We might be able to get away with rectangles most of the time.
Another option is reducing the granularity of heights that are possible, which should result in fewer changes in height per second, and so fewer changes in animations.

I also thought of your idea of having pngs rather than polygons. If there are 16 heights then it's only 256 different pngs. That's assuming water is hidden behind ramps.

I'm not sure how much difference these things would make, but there are my thoughts.

**Comment by colonelfazackerley** [comment]
it drew 4,427 polygons while I left the profiler running. i doubt rectangles are any better.

things like openJDK vs orasun and graphics card drivers might also help. though, perhaps we should just stick to sprites if that is what works out of the box.

**Comment by andybalaam** [comment]
@tttppp wrote:

> Is make rungui the right way to run the game, or is there some other method to avoid debug processing?

Yes, that runs normally, as does runrabbit (https://github.com/andybalaam/rabbit-escape/issues/239)

**Comment by andybalaam** [comment]
Your concerns about performance are definitely valid.  I'd really like to have Rabbit Escape running well on Raspberry Pi (#96) and at the moment (on the Pi 2) it is not really very good - just about playable.  Ideally it would be nice on the Pi 1, but certainly the Pi 2 should be able to cope.

I feel like this isn't doing anything too taxing - there must be a way to make it perform?

**Comment by colonelfazackerley** [comment]
oracle jre 1.8 is much, much better. the fillPolygon calls are so fast that they dont even register.

**Comment by colonelfazackerley** [comment]
The problem is openJDK. 

I installed oracle JRE via repo http://www.duinsoft.nl/packages.php?t=en
and I can switch back to openJDK ``` sudo update-alternatives --config java ```
https://help.ubuntu.com/community/Java

**Comment by colonelfazackerley** [comment]
So, everyone happy if I forge ahead, and those with performance issue use a better VM?

**Comment by tttppp** [comment]
I was trying Oracle Java 7 earlier. The profiler gave me the following:
![java7profile](https://cloud.githubusercontent.com/assets/7073263/13731255/9a147c02-e95c-11e5-8407-bb785809221d.png)
So it looks like the same problem - lots of time spent in drawFilledPoly.

**Comment by tttppp** [comment]
...and like you said, Java 8 runs beautifully!  It's a complete world apart.  Java 7 isn't supported by Oracle any more, so maybe we shouldn't support it either?
![java8profile](https://cloud.githubusercontent.com/assets/7073263/13731329/042455d0-e95e-11e5-83ac-815d8988f575.png)

**Comment by colonelfazackerley** [comment]
we can use the java 7 API, so it can be used, but recommend the oracle JRE 8 for those with performance problems.

**Comment by andybalaam** [comment]
Not keen on requiring Oracle java because it's not properly open source but could probably be persuaded to recommend openjdk 8 if that solves the problem.

**Comment by andybalaam** [comment]
Side note: the Android version we support provides libraries equivalent to java 6 so we can't start using the (very nice) language features of java 8 unless we ditch those users.

**Comment by colonelfazackerley** [comment]
![re](https://cloud.githubusercontent.com/assets/9596719/13812249/ffe56b6e-eb72-11e5-81f1-c59f66845771.gif)
I have rewritten several chunks a few times. got rid of a glitch. glitches remain
- water in a cell with nothing shown in the cell below
- height jump a one point.

it now draws one poly over several cells. i doubt this will improve the performance for openJDK JRE 6. I have a hunch its just a function of the area of the poly.

i should try an opaque poly, see if that can help the old JRE.

I also need to try on android. i have a hunch that it will have better performance for its poly drawing.

**Comment by colonelfazackerley** [comment]
I fixed the right hand side water shown with none below by fixing my falling test and then employing a workaround for #368.

still a couple of glitches.

**Comment by andybalaam** [comment]
Looking better all the time

**Comment by colonelfazackerley** [comment]
I found that opaque polygons are fast too.

I am not suggesting we change from JDK 6 compliance. options for performance
- use oracle JRE 8
- have a config option for swing to have opaque water polygons and use openJDK JRE 7

I have fixed some more glitches. The jumping water height problem remains. its fiddly getting the heights to be continuous at cell boundaries and not jump in time occasionally.

**Comment by colonelfazackerley** [comment]
![re](https://cloud.githubusercontent.com/assets/9596719/13877655/32777abc-ed04-11e5-8f8e-ec54d10c68b1.gif)

slight jumps here, but this seems tolerable. do you guys agree?

**Comment by tttppp** [comment]
Looks much better. I'm happy with that. At some point we can look at fixing the glitches. Do you have a plan for how to handle falling water, eg the cell above the ramp on the right? Maybe some sort of mist might work using the standard animation system?

**Comment by colonelfazackerley** [comment]
I was thinking those wiggly sprites. I am now thinking circular water droplets.

**Comment by colonelfazackerley** [comment]
I also want to add ripples to the tops of the polygons and make the water surface gradient continuous between cells.

**Comment by tttppp** [comment]
Sprites sound good for falling water. The other changes also sound really nice, but I'm not sure they need to be in this change. Once falling water is displayed you'll have something that's a huge improvement on what's already there.

**Comment by andybalaam** [comment]
Looks good enough to be releasable.

**Comment by colonelfazackerley** [comment]
I really must not be distracted from prepping this for merge
https://github.com/andybalaam/rabbit-escape/compare/master...colonelfazackerley:Water3

**Comment by colonelfazackerley** [comment]
Trying to figure out how to fixup all my commits from 23 feb til now into 1.

**Comment by colonelfazackerley** [comment]
After some days I managed to find a way to tidy the commit history. Copy versions of files between branches.

git co otherbranch blah/blah/render/*
and 
git co otherbranch blah/blah/engine/specificfile.java

Then commit them to the new branch.

**Comment by colonelfazackerley** [comment]
suggestions for tests, please. 

https://github.com/andybalaam/rabbit-escape/compare/master...colonelfazackerley:Water4

**Comment by andybalaam** [comment]
`git rebase -i` may have been what you were looking for.

**Comment by colonelfazackerley** [comment]
> git rebase -i may have been what you were looking for. 

Definitely not.

I had commits that needed squashing, then I had merged commits from tttppp. If I simply rebase then the commits from tttppp get changed and it cant be merged to andybalaam/master.

If there is better way, I want to know.

> There are a few tests that check we render correctly by comparing with an actual png image. One or two of those might be appropriate here? 

There are improvements that will change this. I suppose I can change the result too, when I implement them.

**Comment by colonelfazackerley** [comment]
> git rebase -i may have been what you were looking for. 

also, given that I had over 100 commits in the range, the cross branch checkout was much less hassle.

**Comment by tttppp** [comment]
I'm not sure it's less hassle, but you could cherrypick the commits you want (ie the non merges) onto a new branch based on master. If you have a large number of non-merge commits this will be painful.

Alternatively you could checkout the last commit in each non-merge commit range and rebase onto a new branch. Again this sounds fairly painful if there are lots of merge commits.

Possibly the way to avoid this in future is to avoid merging from tttppp/Water, and instead rebase your branch onto the latest tttppp/Water instead. If you'd been doing this then I could have also rebased during development, rather than merging. The downside here is that lots of commits are constantly in flux, so it might be hard to track down new bugs that are introduced during development.

**Comment by tttppp** [comment]
I think this is done now and can be closed. As with most things, there is room for improvement, but I don't think it should be done under this ticket.

**Comment by andybalaam** [comment]
Yep, done!
